File: projects/02-kernelforge/src/common/stats.cpp
#include "common/stats.hpp"

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <stdexcept>

namespace kernelforge {

namespace {
// Linear-interpolation percentile (numpy's default "linear" method), on an
// already-sorted vector. p in [0, 1].
double percentile(const std::vector<double>& sorted, double p) {
  const std::size_t n = sorted.size();
  if (n == 1) return sorted[0];
  const double idx = p * static_cast<double>(n - 1);
  const std::size_t lo = static_cast<std::size_t>(std::floor(idx));
  const std::size_t hi = static_cast<std::size_t>(std::ceil(idx));
  const double frac = idx - static_cast<double>(lo);
  if (lo == hi) return sorted[lo];
  return sorted[lo] + frac * (sorted[hi] - sorted[lo]);
}
} // namespace

TimingStats compute_stats(std::vector<double> samples_ms) {
  if (samples_ms.empty()) {
    throw std::invalid_argument("compute_stats: samples_ms must be non-empty");
  }
  std::sort(samples_ms.begin(), samples_ms.end());

  TimingStats stats;
  stats.min_ms = samples_ms.front();
  stats.max_ms = samples_ms.back();

  double sum = 0.0;
  for (double v : samples_ms) sum += v;
  stats.mean_ms = sum / static_cast<double>(samples_ms.size());

  double sq_diff_sum = 0.0;
  for (double v : samples_ms) {
    const double d = v - stats.mean_ms;
    sq_diff_sum += d * d;
  }
  stats.stddev_ms = samples_ms.size() > 1
                         ? std::sqrt(sq_diff_sum / static_cast<double>(samples_ms.size() - 1))
                         : 0.0;

  stats.median_ms = percentile(samples_ms, 0.5);
  stats.p25_ms = percentile(samples_ms, 0.25);
  stats.p75_ms = percentile(samples_ms, 0.75);
  return stats;
}

} // namespace kernelforge

File: projects/04-flashlite/src/flashlite/stats.py
"""Distribution statistics over raw timing samples (mirrors KernelForge's
src/common/stats.hpp/.cpp exactly, including the linear-interpolation
percentile method -- numpy's default "linear" method -- so results are
directly comparable across the two projects' benchmark JSON).
"""

from __future__ import annotations

import math
from dataclasses import dataclass


@dataclass
class TimingStats:
    min_ms: float
    max_ms: float
    mean_ms: float
    median_ms: float
    stddev_ms: float
    p25_ms: float
    p75_ms: float


def _percentile(sorted_samples: list[float], p: float) -> float:
    n = len(sorted_samples)
    if n == 1:
        return sorted_samples[0]
    idx = p * (n - 1)
    lo = math.floor(idx)
    hi = math.ceil(idx)
    frac = idx - lo
    if lo == hi:
        return sorted_samples[lo]
    return sorted_samples[lo] + frac * (sorted_samples[hi] - sorted_samples[lo])


def compute_stats(samples_ms: list[float]) -> TimingStats:
    if not samples_ms:
        raise ValueError("compute_stats: samples_ms must be non-empty")

    s = sorted(samples_ms)
    n = len(s)
    mean = sum(s) / n
    if n > 1:
        variance = sum((v - mean) ** 2 for v in s) / (n - 1)
        stddev = math.sqrt(variance)
    else:
        stddev = 0.0

    return TimingStats(
        min_ms=s[0],
        max_ms=s[-1],
        mean_ms=mean,
        median_ms=_percentile(s, 0.5),
        stddev_ms=stddev,
        p25_ms=_percentile(s, 0.25),
        p75_ms=_percentile(s, 0.75),
    )
